### PR TITLE
[Relay][Transform] Support Dumping IR to help debugging

### PR DESCRIPTION
**Pull request by zhiics**
This PR enables users to provide the interested passes for debugging during optimization. For example, users can provide a certain pass name for dumping. The module IR will be dumped after this passes is applied.

An example could be like the following:
```python
    shape = (1, 2, 3)
    t = relay.TensorType(shape, "float32")
    x = relay.var("x", t)
    y = relay.add(x, x)
    y = relay.multiply(y, relay.const(2, "float32"))
    func = relay.Function([x], y)

    seq = _transform.Sequential([
        relay.transform.InferType(),
        relay.transform.FoldConstant(),
        relay.transform.DebugPrint(),
        relay.transform.DeadCodeElimination()
    ])

    mod = relay.Module({"main": func})
    with relay.build_config(opt_level=3):
        mod = seq(mod)
```
The output:
```python
Dumping the module IR:
v0.0.1
def @main(%x: Tensor[(1, 2, 3), float32]) -> Tensor[(1, 2, 3), float32] {
  %0 = add(%x, %x) /* ty=Tensor[(1, 2, 3), float32] */
  multiply(%0, 2f /* ty=float32 */) /* ty=Tensor[(1, 2, 3), float32] */
}
```

cc @jroesch @tqchen @icemelon9 @wweic @slyubomirsky @MarisaKirisame

### CODE REVIEW

**Comment by tqchen** [comment]
I am not sure if it is the best approach user would like. For example, if I were to construct the pipeline pragmatically, I might prefer

```python
   seq = _transform.Sequential([
        relay.transform.InferType(),
        relay.transform.FoldConstant(),
        relay.transform.DebugPrint(), 
        relay.transform.DeadCodeElimination()
    ])

```

**Comment by zhiics** [comment]
@tqchen yeah, I also thought about the way you were mentioning, but I decided to implement it this way because 1) I think the major compilation pipeline is usually invisible users, i.e. they will probably only access the compilation using `build_config` without configuring `Sequential` by themselves. Therefore, they only need to debug a pass directly by passing the name of the pass. 2) I made it more like the popular compilers like Clang (clang -debug-pass=xx) and GHC (-ddump-xx).

**Comment by tqchen** [comment]
major compiler pipeline are invisible to the users is because compiler pipeline themselves are not composed pragmatically, but pre configured, and it was really hard to do so in CLI. I do believe as part of python API it is good to use the pragmatic approach as Pytorch did.

We could also provide snippet of the default pipeline so a user can modify pragmatically in python

**Comment by tqchen** [comment]
I think the fundamental question here is whether or not do we want users to pragmatically customize optimization pipelines, do we want to make that part more flexible. One could argue that, like traditional compilers, most pipelines are pre-configured officially and do not need to be tweaked. 

On the other hand, we could also say that given the different kinds of optimizations that we want to perform, there is a bigger amount of demand to pragmatically construct pipelines, and makes users' life easier to do so. In that case, we might want to go beyond the API offered by traditional compilers. 

This was just like the argument of deep learning framework configuration evolution, the first stage config file is just like the options in traditional compilers, then we have TF/MX/Pytorch that offers more pragmatic constructions. 

Would the same thing happen in our case? I do not know, but at least I personally think we should strive to enable that possibility. And learn more from deep learning frameworks like Pytorch.

**Comment by eqy** [comment]
Hooking something in to print every time a pass is run might be too cumbersome to use in some situations, such as when `FoldConstant` is called many times in another pass. I think "show me what happens after this pass" is a more common pattern than "dump every time this pass is run."

**Comment by kevinthesun** [comment]
@tqchen I think it's helpful to encourage user to at least be able to customize(for now just turn on/off?) certain optimization related passes, such as CombineParallelConv2d. The usefulness of these passes can be highly related to input models, and some experiments might be beneficial.

**Comment by zhiics** [comment]
Thanks for all the discussion and feedback:)  I totally agree that it is fundamentally more about if the programmatic or the default compilation is more common used. This is something I am not sure as well. I am very glad to hear the voice from the community.

If we go the programmatic way, I think we probably need to bring in the python pipeline back because other users may need intrusively insert `DebugPrint` pass in the C++ backend and recompile. 

Another thing we probably need to be aware of is that if we only use `opt_level` to control the compilation without baking in the `transform.xx` in the `build_module` in the future. It might not be quite convenient to insert `DebugPrint` after a certain pass.

Any suggestions/ideas about this?

**Comment by tqchen** [comment]
Re the question of whether pipeline constructions should be made in python(will there be duplications if we do things both in python and c++).

I think it boils down to the fact that whether we treat compilation pipelines just like models in deep learning frameworks. Models are implemented in many languages, you see resnet pipelines being implemented in official python part, c++, or other parts of the language. While it is important to have a default one, and if that default one is concise enough, I think it might make sense to bring a python version of pipeline construction.

**Comment by zhiics** [comment]
Okay. I will use the `DebugPrint`. Whether or not bringing the Python pipeline back would be a separate issue. 

Should we allow users to dump all passes instead of inserting `DebugPrint` after each individual passes? If so, should we provide a compilation option for it?

**Comment by tqchen** [comment]
Let us not enable global printing for now and think about the need later

**Comment by zhiics** [comment]
Thanks all. I updated the PR.

File: include/tvm/relay/transform.h [hunk 1]
```diff
@@ -540,6 +540,13 @@ TVM_DLL Pass CanonicalizeCast();
  */
 TVM_DLL Pass EtaExpand();
 
+/*!
+ * \brief Print the IR for a module to help debugging.
+ *
+ * \return the pass.
+ */
+TVM_DLL Pass DebugPrint();
```
Comment (tqchen) [inline include/tvm/relay/transform.h]: How about re-naming it to PrintIR?

Comment (zhiics) [reply include/tvm/relay/transform.h]: Done.

**Comment by slyubomirsky** [comment]
Glad to see the test of the printing. I hope we will have tests for everything that produces error messages or text (perhaps down the line)

**Comment by tqchen** [comment]
THANKS, @slyubomirsky @zhiics @eqy @jroesch this PR is now merged

**Pull request merged**